File: LeetCode/LeetCode Practice Problems/1329- Sort the Matrix Diagonally.cpp
class Solution {
public:
    vector<vector<int>> diagonalSort(vector<vector<int>>& mat) {
        int n = mat.size();
        int m = mat[0].size();
        
        vector<vector<int>> a(n , vector<int> (m , 0));
        
        for(int k=0;k<m;k++) {
            int row = 0;
            int col = k;
            
            vector<int> b;
            while(row<n && col<m) {
                b.push_back(mat[row][col]);
                row++;
                col++;
            }
            
            sort(b.begin(),b.end());
            int i = 0;
            row = 0 , col = k;
            while(row<n && col<m) {
                a[row][col] = b[i++];
                row++;
                col++;
            }
        }
        
        for(int k=1;k<n;k++) {
            int row = k;
            int col = 0;
            
            vector<int> b;
            while(row<n && col<m) {
                b.push_back(mat[row][col]);
                row++;
                col++;
            }
            
            sort(b.begin(),b.end());
            int i = 0;
            row = k , col = 0;
            while(row<n && col<m) {
                a[row][col] = b[i++];
                row++;
                col++;
            }
        }
        
        return a;
    }
};

File: LeetCode/LeetCode Practice Problems/1329-Sort the Matrix Diagonally.java
class Solution {
    
    public int[][] diagonalSort(int[][] a) {
        int n = a.length;
        int m = a[0].length;
        
        for(int j=0;j<m;j++) {
            
            int r = 0 , c = j;
            List<Integer> b = new ArrayList();
            int idx = 0;
            
            while(r<n && c<m) {
                b.add(a[r][c]);
                r++;
                c++;
            }
            
            Collections.sort(b);
            r = 0;
            c = j;
            
            while(r<n && c<m) {
                a[r][c] = b.get(idx);
                idx++;
                r++;
                c++;
            }
        }
        
        for(int i=1;i<n;i++) {
            int r = i , c = 0;
            List<Integer> b = new ArrayList();
            int idx = 0;
            
            while(r<n && c<m) {
                b.add(a[r][c]);
                r++;
                c++;
            }
            
            Collections.sort(b);
            r = i;
            c = 0;
            
            while(r<n && c<m) {
                a[r][c] = b.get(idx);
                idx++;
                r++;
                c++;
            }
        }
        
        return a;
    }
}
